onnxruntime/core/optimizer/attention_fusion_helper.h

```diff
@@ -776,11 +933,29 @@ bool CheckNodesInPathV(const Graph& graph, const Node& reshape, const Node& tran
   head_size = v_reshape_shape[3];
 
   // Check reshape for attention output has shape input (0, 0, -1) or (0, 0, N*H)
+  // In DistilBert, the reshape after qkv paths can not be fused during reshape fusion, so we do not have the correspondig
+  // initializer. We need to get the shape information from the input of concat.
   std::vector<int64_t> reshape_shape;
-  if (!optimizer_utils::AppendTensorFromInitializer(graph, *(reshape.InputDefs()[1]), reshape_shape) ||
-      reshape_shape.size() != 3 ||
+  if (!optimizer_utils::AppendTensorFromInitializer(graph, *(reshape.InputDefs()[1]), reshape_shape)) {
+    const Node* p_concat = graph_utils::GetInputNode(reshape, 1);
+    if (p_concat == nullptr || (*p_concat).OpType() != "Concat")
+      return false;
+    if ((*p_concat).InputDefs().size() != 3)
+      return false;
+    for (size_t input_arg_id = 0; input_arg_id < (*p_concat).InputDefs().size(); ++input_arg_id) {
+      const NodeArg& concat_input_arg_i = *((*p_concat).InputDefs()[input_arg_id]);
+      std::vector<int64_t> shape;
+      if (!optimizer_utils::AppendTensorFromInitializer(graph, concat_input_arg_i, shape)) {
+        reshape_shape.push_back(0);
```

Comment:
It is tricky to have the logic like this here. It might be better to right after matching the shape subgraph.